feat: Improve WebGL support for Logs / RUM

fuzzybinary opened the pull request:
### What and why?

This contains a few improvements to WebGL target support for Unity:

#### Tags on Loggers for RUM

You can now add and remove tags on specific loggers using `DdLogger.AddTag` and `DdLogger.RemoveTag`. Integration tests for web have been updated.

#### Support for "non-custom" actions

All actions in WebGL were reported as `custom`, regardless of what type you specified in `AddAction`. This has been fixed and the correct action type should be reported.

#### Support manual resource tracking

New support for calling `StartResource` and `StopResource*` now report those resources.

#### Support automatic resource tracking

The Browser SDK already reported requests made with UnityWebRequest, but didn't correctly determine first party hosts given in the settings. This has been fixed. Additionally, users using `DatadogTrackedWebRequest` would have resources double reported with the new manual resource tracking support. It has been fixed so that `DatadogTrackedWebRequest` is a passthrough on WebGL builds.

#### Technical note

We are relying on timestamps set in Attribute collections for Web, and extracting them in the JavaScript layer for two reasons. One is because the timestamp is inserted into all of the messages already, so it was a good place to capture the information. The second is because the timestamps need to be 64-bit integers, as they are Milliseconds since Epoch.  

So far as I can determine, there is no easy way to send a `long` to JavaScript (maybe through WebAssembly.BigInt?) but we can serialize it to JSON and deserialize it to a JavaScript integer just fine.

### Review checklist

- [ ] This pull request has appropriate unit and / or integration tests

fuzzybinary committed:
feat(web): Add tags on loggers.

fuzzybinary committed:
feat(web): Support Actions with types other than `custom`

This uses the new `addEvent` function in the Browser SDK's plugin system, and sets up using it for resources as well.

fuzzybinary committed:
chore: Add tracked web request integration testing.

This checks that we are correctly gathering tracked web requests, specifically tested on Web.

fuzzybinary committed:
Add manual resource tracking for Web

awforsythe commented on packages/Datadog.Unity/Plugins/WebGL/DatadogRum.jslib:
```diff
@@ -9,8 +9,41 @@ let ddRumLib = {
             return false;
         }
 
+        let self = this;
+        this._rumPlugin = {
+            name: 'DatadogUnityWeb',
+            onRumStart: function(options) {
+                self.addEvent = options.addEvent;
+            }
+        };
```
OK, so when you've `<script/>`-included `browser-rum` in your web page (or your Unity Web Player page), it defines a global `DD_RUM` object in that document's JavaScript context. That object is the Browser SDK's public RUM interface.

Meanwhile, `ddRumLib` defines the interface that the Unity SDK can use to make calls to the Browser SDK, via Unity Web Player's built-in `.jslib`/`DllImport` interop mechanism. It effectively wraps `DD_RUM` in an interface that can be used from C#.

So in these lines here, for example, we're defining a _plugin_ that will be passed to `DD_RUM.init()` via `config.plugins`, and presumably there's some well-defined interface for the shape of a "plugin" (or, for that matter, a "config" object).

Do we pull in any kind of type definitions from the Browser SDK for validation, or are we operating fully in YOLO-JavaScript-land when we use interfaces like these?

_(This note isn't really PR feedback; just trying to get my bearings.)_

fuzzybinary replied:
That's essentially all correct yes.

We don't, pull in any type definitions from the Browser SDK, but they do exist. We could (and maybe should) research if there's a good way to support TypeScript for our `.jslib`, but I dive that deep for this task.

awforsythe commented on packages/Datadog.Unity/Plugins/WebGL/DatadogRum.jslib:
```diff
@@ -9,8 +9,41 @@ let ddRumLib = {
+                this._navigationStart = window.performance.timing.navigationStart;
+            }
+
+            if (!this._navigationStart) {
+                return timestampMs;
+            }
+            return (timestampMs - this._navigationStart);
```
This logic implies that a timestamp can either be absolute or relative, depending on whether `window.performance.timing.navigationStart` was defined (and nonzero). The fact that `getEventRelativeTime` might return an _absolute_ time seems slightly at odds with the name of the function.

Is it safe and well-defined to fall back to an absolute timestamp, or is it possible that the backend will treat an unmodified `timestampMs` value like `1756820852661` as a relative timestamp and end up thinking that your session has lasted for 55 years?

fuzzybinary replied:
This is a good question -- I'll follow up with @BenoitZugmeyer about it.

BenoitZugmeyer replied:
> Is it safe and well-defined to fall back to an absolute timestamp,

It is not, but at the same time if `navigationStart` isn't defined the Browser SDK won't behave correctly either. So this is probably fine, although I agree that this logic implying that `navigationStart` might not be defined is confusing. Assuming that it is defined would probably be better.

Side note 1: `navigationStart` is deprecated so you might want to use [`performance.timeOrigin`](https://developer.mozilla.org/en-US/docs/Web/API/Performance/timeOrigin) and only fallback to `performance.timing.navigationStart` if it's not defined.

Side note 2: instead of creating a relative time based on the device wall clock, would it be possible to access the browser monothonic clock (`performance.now()`)? It would be more accurate for getting Event context in various "histories" the Browser SDK is maintaining.

fuzzybinary replied:
> Side note 1: navigationStart is deprecated so you might want to use [performance.timeOrigin](https://developer.mozilla.org/en-US/docs/Web/API/Performance/timeOrigin) and only fallback to performance.timing.navigationStart if it's not defined.

I'll add it.

> Side note 2: instead of creating a relative time based on the device wall clock, would it be possible to access the browser monothonic clock (performance.now())? It would be more accurate for getting Event context in various "histories" the Browser SDK is maintaining.

Potentially, but not easily at the moment. We do use a "DateProvider" in C# that we could override, but we'd have to be able to convert `performance.now()` to a C# date time to be able to use it.

Long term, if we move away from the `addEvent` interface, we can let the Browser SDK track all of the timing, which might be prefrable.

BenoitZugmeyer replied:
Sounds good!

fuzzybinary replied:
@BenoitZugmeyer Can your review my changes to get `performance.timeOrigin` in?

BenoitZugmeyer replied:
Looks good ✅

awforsythe commented on packages/Datadog.Unity/Plugins/WebGL/DatadogRum.jslib:
```diff
@@ -26,35 +59,133 @@ let ddRumLib = {
+    DDRum_AddResourceError: function(rawMethod, rawUrl, rawErrorKind, rawErrorMessage, rawErrorStackTrace, rawAttributes) {
+        let id = crypto.randomUUID();
+        let attributesStr = UTF8ToString(rawAttributes);
+        let attributes = JSON.parse(attributesStr) ?? {};
```
More how-do-we-handle-JavaScript questions: does this source get run through transpiler/bundler machinery, or does it end up getting loaded into the Unity Web Player page more-or-less unchanged?

Are ES2020 features like the nullish-coalescing operator (`??`) kosher, or do we need to stick with the more conservative `||`?

fuzzybinary replied:
It's run through emscripten. Unity 2021.2 uses 2.0.19.6 which I think might support ES2020, but I can't find any information on it.

awforsythe reviewed (dismissed):
Looks pretty solid!

fuzzybinary committed:
Fix new conditions for TrackedWebRequestTest

fuzzybinary committed:
Use non-deprecated timeOrigin

Use `window.performance.timeOrigin` if it's available over `window.performance.timing.navigationStart`
